packages/embark-blockchain-process/src/baseBlockchainPlugin.ts

```diff
@@ -0,0 +1,100 @@
+import { Embark, Events } from "embark";
+import Web3 from "web3";
+import { BaseBlockchainClientOptions } from "../index";
+
+const NOT_IMPLEMENTED_EXCEPTION = "This method has not been implemented";
+
+/**
+ * Base blockchain plugin. Extend this class in your custom blockchain client plugin.
+ */
+export class BaseBlockchainPlugin {
+  protected embark: Embark;
+  protected events: Events;
+  protected web3?: Web3;
+  protected blockchainClientOptions: BaseBlockchainClientOptions;
+  protected blockchainConfig: any;
+  /**
+   * Creates an instance of base blockchain plugin.
+   * @param {Embark} embark - Embark object, used to get the Events object.
+   * @param {BaseBlockchainClientOptions} blockchainClientOptions - options for the plugin name
+   * and version, and options that will be passed to the blockchain client.
+   */
+  constructor(embark: Embark, blockchainClientOptions: BaseBlockchainClientOptions) {
+    this.embark = embark;
+    this.events = embark.events;
+    this.blockchainClientOptions = blockchainClientOptions;
+
+    // gets hydrated blockchain config from embark, use it to init
+    this.events.once("config:load:blockchain", (blockchainConfig: any) => {
+      this.blockchainConfig = blockchainConfig;
+      // Only register the blockchain plugin and event actions if the current
+      // run time environment is using this blockchain plugin.
+      // Without this, registered and unfired event actions would cause the
+      // blockchain process to hang.
+      if (this.shouldInit(blockchainConfig)) {
+        this.registerBlockchain();
+        this.registerProviderReadyAction();
+      }
+    });
+  }
+
+  /**
+   * Registers provider ready action. This will be triggered by the blockchain connector
+   * and once registered, the blockchain connector will wait for these actions to
+   * complete before continuing.
+   *
+   * These actions occur prior to the "blockchain:ready" event and provide the perfect
+   * time to create and unlock accounts.
+   */
+  private registerProviderReadyAction() {
+    this.embark.registerActionForEvent("blockchain:provider:ready", async (cb) => {
+      if (!this.createAndUnlockAccounts) {
```

Comment:
This `if` will never trigger, no?
Since it will always be defined here in the Base Class

Reply:
True, this should be removed.